README.md
# vegaview

A [webviewhs](https://github.com/lettier/webviewhs) based plot viewer for
[hvega](https://github.com/DougBurke/hvega).

## Example

```haskell
import qualified Graphics.Vega.VegaLite.View   as Plt
import qualified Graphics.Vega.VegaLite.Simple as Plt

main :: IO ()
main = do
    Plt.show' $ Plt.barChart xLabel yLabel columns freqs
  where
    xLabel  = "foo"
    yLabel  = "bar"
    columns = ["cat", "dog", "horse"]
    freqs   = [0.4, 0.8, 0.5]
```


CHANGELOG.md
# Changelog for `vegaview`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Graphics/Vega/VegaLite/View.hs
{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE OverloadedStrings #-}

-- | VegaList Plot Viewer
module Graphics.Vega.VegaLite.View ( show'
                                   ) where

import           System.Directory          (removeFile)
import           System.IO.Temp            (emptySystemTempFile)
import           System.Posix.User         (getLoginName)
import qualified Graphics.UI.Webviewhs  as WHS
import           Graphics.Vega.VegaLite
import qualified Data.Text              as TXT

-- | Takes path to vega plot and opens it in webview
openPlotWindow :: FilePath -> IO ()
openPlotWindow path = WHS.createWindowAndBlock params
  where
    uri    = TXT.pack $ "file://" ++ path
    params = WHS.WindowParams { WHS.windowParamsTitle      = "VegaLite Viewer"
                              , WHS.windowParamsUri        = uri
                              , WHS.windowParamsWidth      = 800
                              , WHS.windowParamsHeight     = 600
                              , WHS.windowParamsResizable  = True
                              , WHS.windowParamsDebuggable = True }

-- | Plot VegaLite with given Title
show' :: VegaLite -> IO ()
show' plt = do
    usr <- getLoginName
    tmp <- emptySystemTempFile (usr ++ "-vega") 
    toHtmlFile tmp plt >> openPlotWindow tmp >> removeFile tmp


src/Graphics/Vega/VegaLite/Simple.hs
{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Simple Plot functions for convenience
module Graphics.Vega.VegaLite.Simple ( heatmap
                                     , barChart
                                     , barChartErr
                                     , densityPlot
                                     ) where

import           Graphics.Vega.VegaLite as VL
import           Data.Aeson                     (object)
import qualified Data.Text              as TXT

-- | Heatmap with given x and y columns.
heatmap ::  ([DataColumn] -> [DataColumn]) -> TXT.Text -> TXT.Text -> TXT.Text
        -> [Double] -> VegaLite
heatmap colIds xAx yAx zAx grd = toVegaLite [ dat [] , enc [] , cfg [] , mrk [] ]
  where
    dat = dataFromColumns [] . colIds . dataColumn zAx (Numbers   grd)
    cfg = configure . configuration (Axis [Grid True, TickBand BExtent])
    enc = encoding . position X [ PName xAx, PmType Nominal ] 
                   . position Y [ PName yAx, PmType Ordinal ]
                   . color      [ MName zAx, MAggregate Mean ]
    mrk = mark Rect

-- | Draw a bar chart
barChart :: TXT.Text -> TXT.Text -> [TXT.Text] -> [Double] -> VegaLite
barChart xLabel yLabel cols vec = toVegaLite [ dat [], enc [], cfg [], mrk [] ]
  where
    dat = dataFromColumns [] . dataColumn xLabel (Strings cols) 
                             . dataColumn yLabel (Numbers vec)
    enc = encoding . position X [ PName xLabel, PmType Nominal ] 
                   . position Y [ PName yLabel, PmType Quantitative ]
    cfg = configure . configuration (Axis [ Grid True
                                          , TickBand BExtent
                                          , LabelAngle 270.0
                                          ])
    mrk = mark Bar

-- | Draw a bar chart with error bar
barChartErr :: TXT.Text -> TXT.Text -> [TXT.Text] -> [Double] -> Double -> VegaLite
barChartErr xLabel yLabel cols vec err = toVegaLite [ layers ]
  where
    dat    = dataFromColumns [] . dataColumn xLabel (Strings cols)
                                . dataColumn "Value" (Numbers vec)
    enc    = encoding . position X [ PName xLabel, PmType Ordinal ]
                      . position Y [ PName "Value", PmType Quantitative ]
    expr   = TXT.append "datum.Value >= " (TXT.pack . show $ err)
    enc'   = encoding . position X  [ PName xLabel, PmType Ordinal ]
                      . position Y  [ PName "baseline", PmType Quantitative
                                    , PTitle yLabel ]
                      . position Y2 [ PName "Value" ]
                      . color       [ MString "#e45755" ]
    trf    = transform . VL.filter (FExpr expr)
                       . calculateAs (TXT.pack . show $ err) "baseline"
    mrk    = mark Bar
    lay    = layer [ asSpec [ mrk [], enc [] ]
                   , asSpec [ mrk [], trf [], enc' [] ] ]
    jsn    = object []
    dat'   = dataFromJson jsn
    enc''  = encoding . position Y  [ PDatum (Number err) ]
    mrk'   = mark Rule
    mrk''  = mark Text
    lay'   = layer [ asSpec [ mrk' [] ]
                   , asSpec [ mrk'' [ MAlign AlignRight
                                    , MBaseline AlignBottom
                                    , MText "Critical"
                                    , MdX (-2.0)
                                    , MdY (-2.0)
                                    , MXWidth
                                    ] ] ]
    layers = layer [ asSpec [ dat [], lay ]
                   , asSpec [ dat' [], enc'' [], lay' ] ]

-- | Draw Density Plot
densityPlot :: TXT.Text -> [Double] -> VegaLite
densityPlot xLabel vec = toVegaLite [ dat [], trf [], enc [], mrk [] ]
  where
    dat = dataFromColumns [] . dataColumn xLabel (Strings [xLabel])
                             . dataColumn "value" (Numbers vec)
    trf = transform . density "value" [ DnBandwidth 0.3 ]
    enc = encoding . position X [ PName "value", PTitle xLabel
                                , PmType Quantitative ]
                   . position Y [ PName "density", PmType Quantitative ]
    mrk = mark Area


app/Main.hs
{-# OPTIONS_GHC -Wall #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE BlockArguments #-}

module Main (main) where

-- import qualified Graphics.UI.Webviewhs as WHS
import qualified Graphics.Vega.VegaLite.View   as Plt
import qualified Graphics.Vega.VegaLite.Simple as Plt

main :: IO ()
main = do
    Plt.show' $ Plt.barChart xLabel yLabel columns freqs
  where
    xLabel  = "foo"
    yLabel  = "bar"
    columns = ["cat", "dog", "horse"]
    freqs   = [0.4, 0.8, 0.5]

-- main :: IO ()
-- main = do
--     putStrLn "BEFORE"
--     WHS.createWindowAndBlock params
--     putStrLn "AFTER"
--   where
--     params = WHS.WindowParams { WHS.windowParamsTitle      = "PLOT"
--                               , WHS.windowParamsUri        = "https://augustunderground.github.io/vegaview"
--                               , WHS.windowParamsWidth      = 800
--                               , WHS.windowParamsHeight     = 600
--                               , WHS.windowParamsResizable  = True
--                               , WHS.windowParamsDebuggable = True }
